Add zapier doc (#11372)
pingcap/docs: tidb-cloud/integrate-tidbcloud-with-zapier.md

### Actions

The following table lists the actions supported by TiDB Cloud App. Note that some actions need extra resources and you need to prepare the corresponding resources before using the action.

| Action | Description | Resource |
|---|---|---|
| Find Cluster | Finds an existing Serverless tier or Dedicated tier. | None |
| Create Cluster | Creates a new cluster. Only supports creating a free Serverless Tier cluster. | None |
| Find Database | Finds an existing database. | A Serverless Tier cluster |
| Create Database | Creates a new database. | A Serverless Tier cluster |
| Find Table | Finds an existing Table. | A Serverless Tier cluster and a database |
| Create Table | Creates a new table. | A Serverless Tier cluster and a database |
| Create Row | Creates a new row. | A Serverless Tier cluster, a database, and a table |
| Update Row | Updates an existing row. | A Serverless Tier cluster, a database, and a table |
| Find Row | Finds a row in a table via a lookup column. | A Serverless Tier cluster, a database, and a table |
| Find Row (Custom Query) | Finds a row in a table via a custom query the you provide. | A Serverless Tier cluster, a database, and a table |

> 

Reviewer: @shiyuhang0 I made a small change to the table. PTAL.

Author: get

The `New Row` trigger also uses a flexible strategy to generate the `id` field to do the deduplication. The trigger generates the `id` field in the following order:

1. If the result contains an `id` column, use the `id` column.
2. If you specify a `Dedupe Key` in the trigger configuration, use the `Dedupe Key`.
3. If the table has a primary key, use the primary key. If there are multiple primary keys, use the first column.
4. If the table has a unique key, use the unique key.
5. Use the first column of the table.

**New Row (Custom Query) Trigger**

The `New Row (Custom Query)` trigger limits 1,000,000 results in every fetch. 1,000,000 is a large number, and it is only set so as to protect the whole system. It is recommended that your query includes `ORDER BY` and `LIMIT`.

To perform deduplication, your query results must have a unique id field. Otherwise, you will get the `You must return the results with id field` error.

Make sure that your custom query executes in less than 30 seconds. Otherwise, you will get the timeout error.

> ### How to use `find or create` action

Reviewer: ```suggestion
### How do I use the `find or create` action?
```

3. Set up action

   In this step, you need to find a table in your TiDB Cloud cluster. And create a new one if it is not exist.

   Choose the project and cluster from drop-down list. Then you will see the connection information of the database.

   ![img.png](/media/tidb-cloud/zapier/tidbcloud_project.png)

   Enter your password and choose the database from drop-down list. 

   ![img.png](/media/tidb-cloud/zapier/tibdcloud_databse.png)

   The template has set the table name and the DDL which will be executed if the table is not exist. You just need to click **Continue** to enter the next step.

   ![img.png](/media/tidb-cloud/zapier/tibdcloud_ddl.png)

4. Test action

   Click **Test action**, and you will create the table in this step. You can also skip the test and create table when this template is running.

## Step 4: Set up the `Create Row in TiDB Cloud` action

1. Choose app & event

   Keep the default value set by template. Click **Continue** to enter the next step.

2. Choose account

   You can select the account you have created before. Click **Continue** to enter the next step.

   ![img.png](/media/tidb-cloud/zapier/tidbcloud_choose_account.png)

3. Set up action

>    Fill in the `Project Name`, `Cluster Name`, `TiDB password` and `Database Name` just like what you have done in the previous step. Then choose the **github_global_event** table from drop-down list , and you will find the columns of the table will be shown.

Reviewer 2: ```suggestion
   Fill in the `Project Name`, `Cluster Name`, `TiDB Password`, and `Database Name` just like you did in the previous step. Then choose the **github_global_event** table from drop-down list , and you will find the columns of the table will be shown.
```

---
title: Integrate TiDB Cloud with Zapier
summary: Learn how to connect TiDB Cloud to 5000+ Apps with Zapier.
---

# Integrate TiDB Cloud with Zapier

[Zapier](https://zapier.com/app/dashboard) is an automation tool that lets you easily create workflows that involve thousands of apps and services.

Using the [TiDB Cloud app](https://zapier.com/apps/tidb-cloud/integrations) on Zapier enables you to:

- Use TiDB, a MySQl-compatible HTAP database, for free. No need to build locally.
- Make it easier to manage your TiDB Cloud.
- Connect TiDB Cloud to 5000+ apps and automate your workflows.

This guide gives a high-level introduction to the TiDB Cloud app on Zapier and an example of how to use it.

> # Quick start with template

Author: ok

2. Choose account

   Connect to your GitHub account in `Choose account` page. You can create a new account with **Sign in** button or use an existing one. Then click **Continue** to enter the next step.

3. Set up trigger

   The template has done all the things for you in this step. Just click **Continue** to enter the next step.

4. Test trigger

   Click **Test trigger**. It will show the data from GitHub global event if nothing get wrong.

## Step 3: Set up the `Find Table in TiDB Cloud` action

1. Choose app & event

   Keep the default value `Find Table` set by template. Click **Continue** to enter the next step.

2. Choose account

   You will be redirected to a new login page once you click the **Sign in**. Fill in your `Public Key` and `Private Key`. To get API key, follow the [TiDB Cloud API documentation](https://docs.pingcap.com/tidbcloud/api/v1beta#section/Authentication/API-Key-Management). Click **Continue** after you finish.
    
    ![img.png](/media/tidb-cloud/zapier/tidb_findtable_account.png)

3. Set up action

   In this step, you need to find a table in your TiDB Cloud cluster. And create a new one if it is not exist.

>    Choose the project and cluster from drop-down list. Then you will see the connection information of the database.
> 
>    ![img.png](/media/tidb-cloud/zapier/tidbcloud_project.png)
> 
>    Enter your password and choose the database from drop-down list. 
> 
>    ![img.png](/media/tidb-cloud/zapier/tibdcloud_databse.png)
> 
>    The template has set the table name and the DDL which will be executed if the table is not exist. You just need to click **Continue** to enter the next step.
> 
>    ![img.png](/media/tidb-cloud/zapier/tibdcloud_ddl.png)

Reviewer: I updated the filenames because we must ensure all image filenames are **globally unique** across the repository. Please update the image filenames accordingly.

```suggestion
    1. From the drop-down list, choose the project name and cluster name. The connection information of your cluster will be displayed automatically.

        ![Set up project name and cluster name](/media/tidb-cloud/zapier/zapier-set-up-tidbcloud-project-and-cluster.png)

    2. Enter your password.
    
    3. From the drop-down list, choose the database.

        ![Set up database name](/media/tidb-cloud/zapier/zapier-set-up-tidbcloud-databse.png)

        Zapier queries the databases from TiDB Cloud using the password you entered. If no database is found in your cluster, re-enter your password and refresh the page.

    4. In **The table you want to search** box, fill in `github_global_event`. If the table does not exist, the template uses the following DDL to create the table. Click **Continue**.

        ![The create table DDL](/media/tidb-cloud/zapier/zapier-tidb-cloud-create-table-ddl.png)
```

Author: ok

4. Test action

>    Click **Test action**, and you will create the table in this step. You can also skip the test and create table when this template is running.

Reviewer: ```suggestion
   Click **Test action**, and Zapier will create the table. You can also skip the test, and the table will be created when this workflow is running for the first time.
```

## Step 4: Set up the `Create Row in TiDB Cloud` action

1. Choose app & event

>    Keep the default value set by template. Click **Continue** to enter the next step.

Reviewer: ```suggestion
   Keep the default value set by the template. Click **Continue**.
```

2. Choose account

>    You can select the account you have created before. Click **Continue** to enter the next step.

Reviewer: ```suggestion
   Select the account you have chosen when you set up the `Find Table in TiDB Cloud` action. Click **Continue**.
```

   ![img.png](/media/tidb-cloud/zapier/tidbcloud_choose_account.png)

3. Set up action

>    Fill in the `Project Name`, `Cluster Name`, `TiDB Password`, and `Database Name` just like you did in the previous step. Then choose the **github_global_event** table from drop-down list , and you will find the columns of the table will be shown.

Reviewer: ```suggestion
    1. Fill in the **Project Name**, **Cluster Name**, **TiDB Password**, and **Database Name** as in the previous step. 

    2. In the **Table Name**, choose the **github_global_event** table from drop-down list. The columns of the table are displayed.
```

>    ![img.png](/media/tidb-cloud/zapier/tidbcloud_columns.png)

Reviewer: ```suggestion
        ![Table columns](/media/tidb-cloud/zapier/zapier-set-up-tidb-cloud-columns.png)
```

>    Then click the text area, and you will find that you can choose the data from the trigger. Fill in all the columns with trigger's data. Then click **Continue** to enter the next step.

Reviewer: ```suggestion
    3. In the **Columns** box, choose the corresponding data from the trigger. Fill in all the columns, and click **Continue**.
```

>    ![img.png](/media/tidb-cloud/zapier/tidbcloud_triggers_data.png)

Reviewer: ```suggestion
        ![Fill in Columns](/media/tidb-cloud/zapier/zapier-fill-in-tidb-cloud-triggers-data.png)
```

4. Test action

>     Click **Test action** to create a new row in the table. If you check the TiDB, you can find the data

Reviewer: ```suggestion
    Click **Test action** to create a new row in the table. If you check your TiDB Cloud cluster, you can find the data is written successfully.
```

>    ```
>    mysql> select * from test.github_global_event;

Reviewer: ```suggestion
   ```sql
   mysql> SELECT * FROM test.github_global_event;
```

---
title: Integrate TiDB Cloud with Zapier
summary: Learn how to connect TiDB Cloud to 5000+ Apps with Zapier.
---

# Integrate TiDB Cloud with Zapier

[Zapier](https://zapier.com/app/dashboard) is an automation tool that lets you easily create workflows that involve thousands of apps and services.

Use TiDB Cloud App on Zapier enables you:

- Use TiDB, a MySQl-compatible HTAP database for free. No need to build locally.
- Make it easier to manage your TiDB Cloud.
- Connect TiDB Cloud to 5000+ Apps and automate your workflows.

This guide will give a high-level introduction to TiDB Cloud App on Zapier and an example on how to use it.

# Quick start with template

In this section, we will use the preset template as an example to try out TiDB Cloud App on Zapier.

## Prepare

Before you start, you need:

- A [Zapier account](https://zapier.com/app/login).
- A [GitHub account](https://github.com/login).

> - A [TiDB Cloud account](https://tidbcloud.com/signup) and a Serverless Tier on TiDB Cloud, See [TiDB Cloud Quick Start](https://docs.pingcap.com/tidbcloud/tidb-cloud-quickstart#step-1-create-a-tidb-cluster) for more details.

Reviewer: ```suggestion
- A [TiDB Cloud account](https://tidbcloud.com/signup) and a Serverless Tier cluster on TiDB Cloud, See [TiDB Cloud Quick Start](https://docs.pingcap.com/tidbcloud/tidb-cloud-quickstart#step-1-create-a-tidb-cluster) for more details.
```

## Step 1: Get the template

Go to [TiDB Cloud App on Zapier](https://zapier.com/apps/tidb-cloud/integrations) to choose a template and click **Try it**. Then you will enter the editor page. 

> Here we use `Add Github global event into TiDB without custom schema` template as an example. It will add a new row into TiDB Cloud cluster when a new global event is created on GitHub.

Reviewer: I couldn't find this `Add Github global event into TiDB without custom schema` template on Zapier.

Author: The name has been changed. I will change the doc too.

---
title: Integrate TiDB Cloud with Zapier
summary: Learn how to connect TiDB Cloud to 5000+ Apps with zapier.
---

# Integrate TiDB Cloud with zapier

[Zapier](https://zapier.com/app/dashboard) is an automation tool that lets you easily create workflows that involve common apps and services.

Use TiDB Cloud App on zapier enable you connect TiDB Cloud to 5000+ Apps.

This guide describes how to use TiDB Cloud App on zapier.

# Before you start

Make sure you have met the following requires

1. A TiDB Cloud account

> 2. A zapier account

Reviewer 2: Please give links to how to create these accounts.

---
title: Integrate TiDB Cloud with Zapier
summary: Learn how to connect TiDB Cloud to 5000+ Apps with Zapier.
---

# Integrate TiDB Cloud with Zapier

[Zapier](https://zapier.com/app/dashboard) is an automation tool that lets you easily create workflows that involve thousands of apps and services.

Using the [TiDB Cloud app](https://zapier.com/apps/tidb-cloud/integrations) on Zapier enables you to:

> - Use TiDB, a MySQl-compatible HTAP database, for free. No need to build locally.

Reviewer 3: ```suggestion
- Use TiDB, a MySQL-compatible HTAP database, for free. No need to build locally.
```

- [Send emails via Gmail from new custom TiDB queries](https://zapier.com/apps/gmail/integrations/tidb-cloud/1134903/send-emails-via-gmail-from-new-custom-tidb-queries).
- [Add rows to TiDB Cloud from newly caught webhooks](https://zapier.com/apps/tidb-cloud/integrations/webhook/1134955/add-rows-to-tidb-cloud-from-newly-caught-webhooks).
- [Store new Salesforce contacts on TiDB rows](https://zapier.com/apps/salesforce/integrations/tidb-cloud/1134923/store-new-salesforce-contacts-on-tidb-rows).
- [Create TiDB rows for new Gmail emails with resumes and send direct Slack notifications](https://zapier.com/apps/gmail/integrations/slack/1135456/create-tidb-rows-for-new-gmail-emails-with-resumes-and-send-direct-slack-notifications)

## FAQ

### How can I set up the TiDB Cloud account in Zapier?

Zapier requires your **TiDB Cloud API key** to connect with your TiDB Cloud account. Zapier does not need your login account for TiDB Cloud.

To get your TiDB Cloud API key, follow the [TiDB Cloud API documentation](https://docs.pingcap.com/tidbcloud/api/v1beta#section/Authentication/API-Key-Management).

> ### How does TiDB Cloud trigger de-duplication?

Reviewer: ```suggestion
### How TiDB Cloud triggers de-duplication
```

Zapier triggers can work with a polling API call to check for new data periodically (depends on Zapier plan).

TiDB Cloud triggers provide the polling API call which will return a lot of results, most of which Zapier has seen before.

Since we don’t want to trigger an action multiple times when an item in your API exists in multiple distinct polls, we will deduplicate the data with the `id` field.

`New Cluster` and `New Table` can simply use the `cluster_id` and `table_id` as `id` field to do the deduplication. You need not do anything for them. Here I will introduce other triggers.

**New Row Trigger**

First, `New Row` trigger limits 10,000 results in every fetch. This will cause the new rows will not be triggered for they may not be included in this 10000 results.

One way to avoid this is to specify `Order By` configuration in the trigger. For example, once you orderby create time, the new rows will always be included in the 10000 results.

Second, `New Row` will use a flexible strategy to generate the `id` filed to do the deduplication. Here are the priority:

> 1. `id` column if the result contains `id` column.

Reviewer: ```suggestion
1. If the result contains an `id` column, use the `id` column.
```

> 2. `Dedupe Key` if you specify it in the trigger configuration.
> 3. primary key if the table has a primary key (use the first column if there are multiple primary keys).
> 4. unique key if the table has a unique key.
> 5. the first column of the table.

Reviewer: ```suggestion
2. If you specify a `Dedupe Key` in the trigger configuration, use the `Dedupe Key`.
3. If the table has a primary key, use the primary key. If there are multiple primary keys, use the first column.
4. If the table has a unique key, use the unique key.
5. Use the first column of the table.
```

**New Row (Custom Query) Trigger**

> `New Row (Custom Query)` trigger limits 1,000,000 results in every fetch. It is a large number, and we only set it to protect the whole system. So, Your query is desired to include order and limit.

Reviewer: ```suggestion
The `New Row (Custom Query)` trigger limits 1,000,000 results in every fetch. 1,000,000 is a large number, and it is only set so as to protect the whole system. It is recommended that your query includes `ORDER BY` and `LIMIT`.
```

> As for deduplication, your query results must have a unique id field or you will get the `You must return the results with id field` error.

Reviewer: ```suggestion
To perform deduplication, your query results must have a unique id field. Otherwise, you will get the `You must return the results with id field` error.
```

> Note that your custom query must run less than 30 seconds.

Reviewer: 如果超过 30 秒会怎么样

Author: 会报错，我加一下

> ### Resources required by TiDB Cloud actions

Reviewer: 这一部分能否合并到上面 actions 的部分？用一个表格或者列表的形式，让用户在查看 actions 的时候就能看到每个 action 需要的资源

Author: 我尝试一下

> Most of the TiDB Cloud actions require additional resources you need prepare.

Reviewer: ```suggestion
TiDB Cloud actions might require additional resources. You need to prepare the resources as follows.
```

> 1. Actions need an existing TiDB Cloud serverless tier (TiDB database):

Reviewer: ```suggestion
* Actions that require an existing Serverless Tier cluster in TiDB Cloud:
```

   - Find Database: Finds an existing Database.
   - Create Database: Creates a new database.
   - Find Table: Finds an existing Table.
   - Create Table: Creates a new table.
   - Create Row: Creates a new row.
   - Update Row: Updates an existing row.
   - Find Row: Finds a row in a table via a lookup column.
   - Find Row (Custom Query): Finds a Row in a table via a custom query in your control.

> 2. Actions need an existing database in TiDB:

Reviewer: ```suggestion
* Actions that require an existing database in TiDB:
```

   - Find Table: Finds an existing Table.
   - Create Table: Creates a new table.
   - Create Row: Creates a new row.
   - Update Row: Updates an existing row.
   - Find Row: Finds a row in a table via a lookup column.
   - Find Row (Custom Query): Finds a Row in a table via a custom query in your control.

> 3. Actions need an existing table (including the schema) in TiDB:

Reviewer: ```suggestion
* Actions that require an existing table (including the schema) in TiDB:
```

   - Create Row: Creates a new row.
   - Update Row: Updates an existing row.
   - Find Row: Finds a row in a table via a lookup column.
   - Find Row (Custom Query): Finds a Row in a table via a custom query in your control.

### How to use `find or create` action

> `Find or create` action enables you create a resource when it is not exist. Here is an example:

Reviewer: ```suggestion
`Find or create` action enables you to create a resource if it does not exist. Here is an example:
```

1. Choose `Find Table` action

> 2. Click `Create TiDB Cloud Table if it doesn’t exist yet?` button to enable `find and create` in `set up action` step

Reviewer: ```suggestion
2. In the`set up action` step, tick the `Create TiDB Cloud Table if it doesn’t exist yet?` box to enable `find and create`.
```

   ![Find and create](/media/tidb-cloud/zapier/zapier-tidbcloud-find-and-create.png)

> In this example, we will create a table if it does not exist yet. Note that the table will be created directly if you test your action.

Reviewer: ```suggestion
This workflow creates a table if it does not exist yet. Note that the table will be created directly if you test your action.
```
